sexpdata.py

```diff
@@ -462,78 +478,138 @@ class Symbol(SExpBase):
+    def __new__(cls, *args):
+        if not args:
+            raise ValueError("Expected an Iterable/Mapping argument or *args")
+        x = args[0] if len(args) == 1 else args
 
-    def tosexp(self, tosexp=tosexp):
-        return uformat("'{0}", tosexp(self._val))
+        if isinstance(x, Mapping):
+            plist_pairs = ((Symbol(':' + k), v) for k, v in x.items())
+            return tuple.__new__(cls, (chain.from_iterable(plist_pairs),))
+        elif not isinstance(x, basestring) and isinstance(x, Iterable):
+            return tuple.__new__(cls, (x,))
+        else:
+            return tuple.__new__(cls, ((x,),)) # unary *args
 
+    @staticmethod
+    def from_opener(opener, val):
+        cls_map = dict((cls.opener, cls) for cls in Delimiters.__subclasses__())
+        if opener in cls_map.keys():
+            return cls_map[opener](val)
+        else:
+            raise TypeError
 
-class Bracket(SExpBase):
+@tosexp.register(Delimiters)
+def _(self, **kwds):
+    return (self.__class__.opener +
+            ' '.join(tosexp(x, **kwds) for x in self.I) +
+            self.__class__.closer)
 
-    def __init__(self, val, bra):
-        assert bra in BRACKETS  # FIXME: raise an appropriate error
-        super(Bracket, self).__init__(val)
-        self._bra = bra
 
-    def __repr__(self):
-        return uformat("{0}({1!r}, {2!r})",
-            self.__class__.__name__, self._val, self._bra)
+class Brackets(Delimiters):
+    """
+    Outputs an Iterable or Mapping with square brackets.
 
-    def tosexp(self, tosexp=tosexp):
-        bra = self._bra
-        ket = BRACKETS[self._bra]
-        c = ' '.join(tosexp(v) for v in self._val)
-        return uformat("{0}{1}{2}", bra, c, ket)
+    Selectively make a container an array:
+
+    >>> dumps(Brackets(list(range(5))))
+    '[0 1 2 3 4]'
+
+    >>> dumps(Brackets(dict(a=1)))
+    '[:a 1]'
+    """
+
+    opener, closer = '[', ']'
+
+
+class Parens(Delimiters):
+    """
+    Outputs an Iterable or Mapping with parentheses.
+
+    By default Iterables and Mappings output with parentheses.
+
+    >>> dumps(range(5))
+    '(0 1 2 3 4)'
+    >>> dumps(dict(a=1))
+    '(:a 1)'
+
+    Selectively override the tuple_as='array' default parameter:
+
+    >>> dumps((0, Parens((1, 2, 3)), 4), tuple_as='array')
+    '[0 (1 2 3) 4]'
+    """
+
+    opener, closer = '(', ')'
+
+
+class Splice(Delimiters):
+    """
+    Outputs an Iterable or Mapping without delimiters. This acts like a "comma
+    splice", inserting a stream into its parent stream without mutations.
+
+    >>> dumps((0, range(1, 4), 4))
+    '(0 (1 2 3) 4)'
+    >>> dumps((0, Splice(range(1, 4)), 4))
+    '(0 1 2 3 4)'
+
+    This is also useful for emitting a series of S-expressions to a filelike.
+
+    >>> import itertools, sys
+    >>> dump(Splice(itertools.product('ab', range(3))), sys.stdout,
+    ...      str_as='symbol')
+    (a 0) (a 1) (a 2) (b 0) (b 1) (b 2)
+    """
+
+    opener = closer = ''
```

Comment:
Not sure I understand the idea behind `Splice`. Firstly, it is purely serialization-specific. Then the first example
```
dumps((0, Splice(range(1, 4)), 4))
```
is really python-specific and the same should probably be achieved by doing
```
dumps((0, *range(1, 4), 4))
```

The example about dumping multiple objects IMHO should also be treated differently. With Splice it would create an asymmetry: there would be a way to dump multiple objects, but not to read them back. So I guess it would be better to add that functionality via separate functions, like `dump_many` and `load_many` or smth like that, and then `Splice` loses its utility completely.

Reply:
I'm OK with removing Splice.